ctru-rs/src/services/uds.rs

```diff
@@ -0,0 +1,960 @@
+
+        ResultCode(unsafe {
+            ctru_sys::udsCreateNetwork(
+                &network as *const _,
+                passphrase.as_ptr().cast(),
+                passphrase.len(),
+                context.as_mut_ptr(),
+                channel,
+                Self::RECV_BUF_SIZE,
+            )
+        })?;
+
+        let context = unsafe { context.assume_init() };
+
+        self.network.replace(network);
+
+        self.context.replace(context);
+
+        Ok(())
+    }
+
+    /// Destroy the current network.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if no network has been created.
+    /// See [`Uds::create_network()`] to create a network.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # use std::error::Error;
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// #
+    /// use ctru::services::uds::{ConnectionType, SendFlags, Uds};
+    /// let mut uds = Uds::new(None)?;
+    ///
+    /// uds.create_network(b"HBW\x10", None, None, b"udsdemo passphrase c186093cd2652741\0", 1)?;
+    /// uds.destroy_network()?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[doc(alias = "udsDestroyNetwork")]
+    pub fn destroy_network(&mut self) -> Result<(), Error> {
+        if self.service_status() != ServiceStatus::Server {
+            return Err(Error::NoNetwork);
+        }
+
+        // should always be true
+        if self.context.is_some() {
+            self.unbind_context()?;
+        }
+
+        ResultCode(unsafe { ctru_sys::udsDestroyNetwork() })?;
+
+        self.network = None;
+
+        Ok(())
+    }
+
+    /// Set the app data for the currently hosted network.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if no network has been created.
+    /// See [`Uds::create_network()`] to create a network.
+    /// This function will also return an error if the provided buffer is too large (see [`Uds::MAX_APPDATA_SIZE`]).
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # use std::error::Error;
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// #
+    /// use ctru::services::uds::{ConnectionType, SendFlags, Uds};
+    /// let mut uds = Uds::new(None)?;
+    ///
+    /// uds.create_network(b"HBW\x10", None, None, b"udsdemo passphrase c186093cd2652741\0", 1)?;
+    /// uds.set_appdata(b"Test appdata.\0")?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[doc(alias = "udsSetApplicationData")]
+    pub fn set_appdata(&self, data: &[u8]) -> Result<(), Error> {
+        if self.service_status() != ServiceStatus::Server {
+            return Err(Error::NoNetwork);
+        }
+
+        if data.len() > Self::MAX_APPDATA_SIZE {
+            return Err(Error::TooMuchAppData);
+        }
+
+        ResultCode(unsafe { ctru_sys::udsSetApplicationData(data.as_ptr().cast(), data.len()) })?;
```

Comment:
Is the `appdata` saved within the struct? If it just holds the pointer it might cause segmentation faults when the original buffer is dropped.

Reply:
It's not entirely clear from the `libctru` documentation, but reading [the code](https://github.com/devkitPro/libctru/blob/master/libctru/source/services/uds.c#L744), it seems like it's copied to a TLS buffer. Oh god, not TLS again.

Reply:
I mean, it's not like TLS storage doesn't work. The linker just hates us in particular. :laughing: